src/drivers/snapdragon_pwm_out/snapdragon_pwm_out.cpp

```diff
@@ -0,0 +1,608 @@
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include <stdint.h>
+
+#include <px4_tasks.h>
+#include <px4_getopt.h>
+#include <px4_posix.h>
+#include <errno.h>
+#include <cmath>	// NAN
+
+
+#include <uORB/uORB.h>
+#include <uORB/topics/actuator_controls.h>
+#include <uORB/topics/actuator_outputs.h>
+#include <uORB/topics/actuator_armed.h>
+
+#include <drivers/drv_hrt.h>
+#include <drivers/drv_mixer.h>
+#include <systemlib/mixer/mixer.h>
+#include <systemlib/mixer/mixer_load.h>
+#include <systemlib/mixer/mixer_multirotor.generated.h>
+#include <systemlib/param/param.h>
+#include <systemlib/pwm_limit/pwm_limit.h>
+#include <dev_fs_lib_pwm.h>
+
+/*
+ * This driver is supposed to run on Snapdragon.
+ * Author: Dennis Mannhart
+ * email: dennis.mannhart@gmail.com
+ */
+namespace snapdragon_pwm
+{
+
+static px4_task_t _task_handle = -1; // handle to the task main thread
+volatile bool _task_should_exit = false; // should task exit
+static bool _is_running = false; // is app running
+
+static const int NUM_PWM = 4;
+static char _device[64] = "/dev/pwm-1";
+
+// pwm configuration with start pin 27
+static const int PIN_GPIO = 27;
+static struct ::dspal_pwm_ioctl_signal_definition _signal_definition;
+static struct ::dspal_pwm _pwm_gpio[NUM_PWM];
+static struct ::dspal_pwm_ioctl_update_buffer *_update_buffer;
+static struct ::dspal_pwm *_pwm;
+
+
+// ToDo: check frequency required
+static const int FREQUENCY_PWM = 400;
+
+
+/* filename: /dev/fs is mapped to /usr/share/data/adsp */
+static const char *MIXER_FILENAME = "/dev/fs/quad_x.main.mix";
```

Comment:
Is this mixer file already there or copied automatically? Or does it need to be added to the upload script?